### PR TITLE
[camera] android-rework part 5: Android FPS range, resolution and sensor orientation features

### PR DESCRIPTION
**Pull request by mvanbeusekom**
Adds FPS range, resolution and sensor orientation features containing the implementation to handle FPS range, resolution and sensor orientation via the Android Camera2 API.

This is the fifth PR in a series of pull-requests which will gradually introduce changes from PR #3651, making it easier to review the code (as discussed with @stuartmorgan).

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
  - Since this will be a series of pull-requests I will update the version in the last PR of the series.
- [ ] I updated CHANGELOG.md to add a description of the change.
  - Since this will be a series of pull-requests I will update the version in the last PR of the series.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

**Comment by acoutts** [comment]
@googlebot I consent.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        angle = 180;
+        break;
+      case LANDSCAPE_LEFT:
+        angle = 90;
+        break;
+      case LANDSCAPE_RIGHT:
+        angle = 270;
+        break;
+    }
+    if (isFrontFacing) angle *= -1;
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:92]: Braces

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    return (angle + sensorOrientation + 360) % 360;
+  }
+
+  private void startSensorListener() {
+    if (orientationEventListener != null) return;
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:97]: Braces

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+            }
+          }
+        };
+    if (orientationEventListener.canDetectOrientation()) {
+      orientationEventListener.enable();
+    }
+  }
+
+  private void startUIListener() {
+    if (broadcastReceiver != null) return;
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:117]: Braces

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+              }
+            }
+          }
+        };
+    activity.registerReceiver(broadcastReceiver, orientationIntentFilter);
+    broadcastReceiver.onReceive(activity, null);
+  }
+
+  private void stopSensorListener() {
+    if (orientationEventListener == null) return;
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:136]: Braces

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    orientationEventListener.disable();
+    orientationEventListener = null;
+  }
+
+  private void stopUIListener() {
+    if (broadcastReceiver == null) return;
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:142]: Braces

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+      boolean isFrontFacing,
+      int sensorOrientation) {
+    this.activity = activity;
+    this.messenger = messenger;
+    this.isFrontFacing = isFrontFacing;
+    this.sensorOrientation = sensorOrientation;
+  }
+
+  public void start() {
+    startSensorListener();
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:57]: Could you add a comment here explaining why we listen to both? It's not clear to me why we need to compute the orientation two different ways. Is it a case of needing a fallback if one isn't available?

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    } else {
+      return Configuration.ORIENTATION_PORTRAIT;
+    }
+  }
+
+  @SuppressWarnings("deprecation")
+  private Display getDisplay() {
+    return ((WindowManager) activity.getSystemService(Context.WINDOW_SERVICE)).getDefaultDisplay();
+  }
+}
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:214]: This class has a lot of non-trivial logic, but no tests.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/fpsrange/FpsRangeFeature.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+package io.flutter.plugins.camera.features.fpsrange;
+
+import android.hardware.camera2.CaptureRequest;
+import android.util.Log;
+import android.util.Range;
+import io.flutter.plugins.camera.CameraProperties;
+import io.flutter.plugins.camera.features.CameraFeature;
+
+public class FpsRangeFeature extends CameraFeature<Range<Integer>> {
```
[the remark this answers is not in the document]
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/fpsrange/FpsRangeFeature.java:16]: Done.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/fpsrange/FpsRangeFeature.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+          if (upper >= 10) {
+            if (currentSetting == null || upper > currentSetting.getUpper()) {
+              currentSetting = range;
+            }
+          }
+        }
+      }
+    } catch (Exception e) {
+      // TODO: maybe just send a dart error back
+      //            pictureCaptureRequest.error("cameraAccess", e.getMessage(), null);
```
[the remark this answers is not in the document]
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/fpsrange/FpsRangeFeature.java]: After some refactoring the need for error handling is removed. I have removed the obsolete `try...catch` statement.

File: packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/features/fpsrange/FpsRangeFeatureTest.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  @SuppressWarnings("unchecked")
+  public void updateBuilder_should_set_ae_target_fps_range() {
+    CaptureRequest.Builder mockBuilder = mock(CaptureRequest.Builder.class);
+    FpsRangeFeature fpsRangeFeature = createTestInstance();
+
+    fpsRangeFeature.updateBuilder(mockBuilder);
+
+    verify(mockBuilder).set(eq(CaptureRequest.CONTROL_AE_TARGET_FPS_RANGE), any(Range.class));
+  }
+
```
[the remark this answers is not in the document]
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/features/fpsrange/FpsRangeFeatureTest.java:70]: Has become obsolete

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    recordingProfile =
+        getBestAvailableCamcorderProfileForResolutionPreset(cameraName, initialSetting);
+    captureSize = new Size(recordingProfile.videoFrameWidth, recordingProfile.videoFrameHeight);
+
+    previewSize = computeBestPreviewSize(cameraName, initialSetting);
+  }
+
+  public static CamcorderProfile getBestAvailableCamcorderProfileForResolutionPreset(
+      String cameraName, ResolutionPreset preset) {
+    int cameraId = Integer.parseInt(cameraName);
```
[the remark this answers is not in the document]
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java]: Refactored the code to only accept an integer.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java (hunk 6 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:73]: Done.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java (hunk 7 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:307]: Done.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java (hunk 5 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java]: Done.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java (hunk 4 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java]: Done.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java (hunk 3 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java]: Done.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java (hunk 1 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java]: Done.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java (hunk 2 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java]: Done.

**Comment by mvanbeusekom** [comment]
@stuartmorgan thank you for reviewing this PR, I have processed all feedback and would appreciate it if you could have another look.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/fpsrange/FpsRangeFeature.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        int upper = range.getUpper();
+
+        // There is a bug in the Pixel 4A where it cannot support 60fps modes
+        // even though they are reported as supported by
+        // `getControlAutoExposureAvailableTargetFpsRanges`.
+        // For max device compatibility we will keep FPS under 60 even if they report they are
+        // capable of achieving 60 fps.
+        // https://issuetracker.google.com/issues/189237151
+        if (upper >= 10 && upper < 60) {
+          if (currentSetting == null || upper > currentSetting.getUpper()) {
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/fpsrange/FpsRangeFeature.java:40]: I think it would make more sense to flip the nesting of these conditions; the main logic (which is this one) should be the outer loop, and then workaround (the line above and its comment) should be the inner one. As written now, it reads like the whole block is for the workaround, which isn't the case.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/fpsrange/FpsRangeFeature.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+  public FpsRangeFeature(CameraProperties cameraProperties) {
+    super(cameraProperties);
+
+    Range<Integer>[] ranges = cameraProperties.getControlAutoExposureAvailableTargetFpsRanges();
+
+    if (ranges != null) {
+      for (Range<Integer> range : ranges) {
+        int upper = range.getUpper();
+
+        // There is a bug in the Pixel 4A where it cannot support 60fps modes
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/fpsrange/FpsRangeFeature.java:33]: Is there any reasonable way we could gate this workaround to that device, rather than changing the behavior for all Android devices?

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+        } else {
+          throw new IllegalArgumentException(
+              "No capture session available for current capture session.");
+        }
+    }
+  }
+
+  @VisibleForTesting
+  static Size computeBestPreviewSize(int cameraId, ResolutionPreset preset) {
+    if (preset.ordinal() > ResolutionPreset.high.ordinal()) {
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:96]: It would be good to add a comment as to why this is done; IIRC there was a PR that did this originally, and other PRs that have tried to revert it without knowing why it was there.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+   * Gets the optimal preview size based on the configured resolution.
+   *
+   * @return The optimal preview size.
+   */
+  public Size getPreviewSize() {
+    return this.previewSize;
+  }
+
+  /**
+   * Get the optimal capture size based on the configured resolution.
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:160]: Nit: Gets

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+    return new Size(profile.videoFrameWidth, profile.videoFrameHeight);
+  }
+
+  private void configureResolution(ResolutionPreset resolutionPreset, int cameraId) {
+    recordingProfile =
+        getBestAvailableCamcorderProfileForResolutionPreset(cameraId, resolutionPreset);
+    captureSize = new Size(recordingProfile.videoFrameWidth, recordingProfile.videoFrameHeight);
+
+    previewSize = computeBestPreviewSize(cameraId, resolutionPreset);
+  }
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:111]: What's the order of the things in this class (per https://google.github.io/styleguide/javaguide.html#s3.4.2-ordering-class-contents)? I'm not seeing an obvious grouping.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java [hunk 17]
```diff
@@ -0,0 +1,311 @@
...
+
+  private void startSensorListener() {
+    if (orientationEventListener != null) {
+      return;
+    }
+    orientationEventListener =
+        new OrientationEventListener(activity, SensorManager.SENSOR_DELAY_NORMAL) {
+          @Override
+          public void onOrientationChanged(int angle) {
+            if (!isSystemAutoRotationLocked()) {
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:156]: The logic in these handlers isn't getting tested (e.g., if one of these checks were reversed, no tests would catch it). If the testing is going to be via calling private methods rather than mocking the listeners—I'm not sure how feasible that is; I've done very little Java—then as much logic as possible should be in those test-called private methods.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Looking good, just some small things

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/fpsrange/FpsRangeFeature.java (hunk 13 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/fpsrange/FpsRangeFeature.java]: Yes it is possible, I updated the code accordingly

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java (hunk 16 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:110]: Fixed, there is no obvious (or hidden) grouping. I removed the empty line.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java (hunk 15 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java]: Fixed.

**Comment by mvanbeusekom** [comment]
@stuartmorgan I have processed your feedback and would appreciate it if you can have another look. Also I am not sure why the `cla/google` step is failing. The `cla: yes` label is correctly applied since the 8th of April and no other users but me have worked on this PR since.

**Comment by stuartmorgan** [comment]
You landed another co-authored commit after https://github.com/flutter/plugins/pull/3799#issuecomment-815886190, maybe it needs another consent message? I'm not sure exactly how that check is implemented internally.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java (hunk 16 above)
Comment (stuartmorgan) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:121]: The comment here wasn't about the empty line, but the ordering of methods in the class (see link)

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java [hunk 18]
```diff
@@ -30,14 +30,18 @@
    *
    * @param cameraProperties Collection of characteristics for the current camera device.
    * @param resolutionPreset Platform agnostic enum containing resolution information.
-   * @param cameraId Camera identifier of the camera for which to configure the resolution.
+   * @param cameraName Camera identifier of the camera for which to configure the resolution.
    */
   public ResolutionFeature(
-      CameraProperties cameraProperties, ResolutionPreset resolutionPreset, int cameraId) {
+      CameraProperties cameraProperties, ResolutionPreset resolutionPreset, String cameraName) {
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:36]: Why the change from int to string here, if it has to be an int in practice?

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java [hunk 19]
```diff
@@ -129,7 +139,7 @@ public void setValue(ResolutionPreset value) {
   // Always supported
```
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:139]: This is no longer true with the latest change.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java (hunk 18 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:36]: Because Android supplies a list of camera identifies of type `String` and the whole API will work with the `String` type except for the `CamcorderProfile.hasProfile` and `CamcorderProfile.get` which demand the camera identifier to be an `int` (see Android doc [here](https://developer.android.com/reference/android/media/CamcorderProfile)). 

Since the `ResolutionFeature` is the only feature making use of the `CamcorderProfile` SDK we decided to leave the conversion in this class and use the `checkIsSupported` method to indicate if the feature is supported for the selected camera. For some external camera devices the camera identifier will be a `String` and for those the `ResolutionFeature` will not be supported.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java (hunk 19 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java]: Done.

**Comment by mvanbeusekom** [comment]
> You landed another co-authored commit after [#3799 (comment)](https://github.com/flutter/plugins/pull/3799#issuecomment-815886190), maybe it needs another consent message? I'm not sure exactly how that check is implemented internally.

Strange after pushing the changes you requested (removed the obsolete comment) the CLA is happy again. Anyway would appreciate it if you could have another look.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java (hunk 18 above)
Comment (stuartmorgan) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:36]: Makes sense, thanks!

**Review by stuartmorgan — APPROVED** [review]
LGTM

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java (hunk 16 above)
Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:175]: Apologies, I completely missed section 3.4.2 in the article and thought you referred to the "Formatting" section. I see what you mean now and update the order of methods so grouping is more logical.
